**Step 1: Importing Datasets**

In [ ]:
import pandas as pd
orders_df=pd.read_csv("/content/orders DA.csv")

In [ ]:
orders_df.head()

,order_id,customer_id,order_date,order_time,status,city,state,pincode,total_amount,gst_amount,shipping_charge,discount_amount,final_amount,payment_method,shipping_partner,tracking_id,delivered_date,is_cod,channel
0,ORD000001,CUST02946,2025-05-22,17:10:00,Processing,Pune,Maharashtra,300862,100984.19,16364.64,0,20486.45,100984.19,Cash on Delivery,BlueDart,TRK354572763,NaT,1,App
1,ORD000002,CUST04232,2023-10-16,05:32:00,Delivered,Ludhiana,Punjab,387852,254478.01,41599.80,0,19338.79,254478.01,UPI,DTDC,TRK272838570,2023-10-19,0,Website
2,ORD000003,CUST08078,2023-10-08,12:32:00,Cancelled,Meerut,Uttar Pradesh,294014,32557.06,3562.92,0,696.86,32557.06,Debit Card,Amazon Logistics,TRK338505401,NaT,0,App
3,ORD000004,CUST09612,2023-10-02,08:05:00,Delivered,Nashik,Maharashtra,626047,24836.94,3092.88,0,4029.94,24836.94,UPI,DTDC,TRK289792017,2023-10-09,0,Mobile Web
4,ORD000005,CUST03065,2024-06-15,20:19:00,Delivered,Moradabad,Uttar Pradesh,409210,156689.29,25092.36,0,7805.07,156689.29,UPI,Ecom Express,TRK978240041,2024-06-22,0,App


In [ ]:
orders_df=pd.read_csv("/content/orders DA.csv")
customers_df=pd.read_csv("/content/customers DA.csv")
inventory_df=pd.read_csv("/content/inventory DA.csv")
order_items_df=pd.read_csv("/content/order_items DA.csv")
payments_df=pd.read_csv("/content/payments DA.csv")
products_df=pd.read_csv("/content/products DA.csv")
returns_df=pd.read_csv("/content/returns DA.csv")
suppliers_df=pd.read_csv("/content/suppliers DA.csv")
print("All * datasets are loaded successfully")

All * datasets are loaded successfully


In [ ]:
orders_df.shape

(50000, 19)

In [ ]:
orders_df.columns

Index(['order_id', 'customer_id', 'order_date', 'order_time', 'status', 'city',
       'state', 'pincode', 'total_amount', 'gst_amount', 'shipping_charge',
       'discount_amount', 'final_amount', 'payment_method', 'shipping_partner',
       'tracking_id', 'delivered_date', 'is_cod', 'channel'],
      dtype='object')

In [ ]:
order_items_df.shape

(100000, 11)

In [ ]:
order_items_df.columns

Index(['item_id', 'order_id', 'product_id', 'product_name', 'category',
       'quantity', 'unit_price', 'gst_rate', 'gst_amount', 'discount_amount',
       'total_price'],
      dtype='object')

In [ ]:
suppliers_df.shape

(200, 14)

In [ ]:
suppliers_df.columns

Index(['supplier_id', 'supplier_name', 'contact_person', 'email', 'phone',
       'city', 'state', 'pincode', 'category', 'gstin', 'payment_terms_days',
       'rating', 'created_date', 'is_active'],
      dtype='object')

In [ ]:
inventory_df.shape

(1000, 11)

In [ ]:
inventory_df.columns

Index(['inventory_id', 'product_id', 'warehouse_location',
       'quantity_available', 'quantity_reserved', 'reorder_level',
       'reorder_quantity', 'last_restocked_date', 'unit_cost',
       'total_inventory_value', 'status'],
      dtype='object')

In [ ]:
payments_df.shape

(50000, 13)

In [ ]:
payments_df.columns

Index(['payment_id', 'order_id', 'customer_id', 'payment_date', 'payment_time',
       'payment_method', 'amount', 'status', 'transaction_id', 'bank_name',
       'gateway', 'refund_amount', 'refund_date'],
      dtype='object')

In [ ]:
customers_df.shape

(10000, 17)

In [ ]:
customers_df.columns

Index(['customer_id', 'first_name', 'last_name', 'email', 'phone', 'city',
       'state', 'pincode', 'gender', 'age', 'segment', 'registration_date',
       'last_login_date', 'is_verified', 'is_active', 'total_orders',
       'total_spent'],
      dtype='object')

In [ ]:
products_df.shape

(1000, 17)

In [ ]:
products_df.columns

Index(['product_id', 'product_name', 'category', 'subcategory', 'brand', 'sku',
       'mrp', 'selling_price', 'cost_price', 'gst_rate', 'hsn_code',
       'weight_grams', 'supplier_id', 'rating', 'review_count', 'is_active',
       'launch_date'],
      dtype='object')

In [ ]:
returns_df.shape

(10000, 10)

In [ ]:
returns_df.columns

Index(['return_id', 'order_id', 'customer_id', 'return_date', 'reason',
       'return_amount', 'refund_status', 'refund_date', 'return_condition',
       'remarks'],
      dtype='object')

****Step 2: Profiling All8 Datasets****

In [ ]:
datasets={
    'orders':orders_df,
    'order_items':order_items_df,
    'suppliers':suppliers_df,
    'inventory':inventory_df,
    'payments':payments_df,
    'customers':customers_df,
    'products':products_df,
    'returns':returns_df
}
print("  ----Datasets profiling (rows & columns)")

for name,data in datasets.items():
  print(f"{name:<12} has    ||   {data.shape[0]:<7} rows    ||  and {data.shape[1]} columns")

  ----Datasets profiling (rows & columns)
orders       has    ||   50000   rows    ||  and 19 columns
order_items  has    ||   100000  rows    ||  and 11 columns
suppliers    has    ||   200     rows    ||  and 14 columns
inventory    has    ||   1000    rows    ||  and 11 columns
payments     has    ||   50000   rows    ||  and 13 columns
customers    has    ||   10000   rows    ||  and 17 columns
products     has    ||   1000    rows    ||  and 17 columns
returns      has    ||   10000   rows    ||  and 10 columns


In [ ]:
print(orders_df["status"].value_counts())

status
Delivered     32499
Cancelled      5996
Shipped        5071
Returned       3933
Processing     2501
Name: count, dtype: int64


**Step 3: Missing Values and Percentagesfor all8 tabels**

In [ ]:
print("Missing Values Report")
print("-" * 50)

for name, df in datasets.items():
    print(f"\n{name}")
    print(df.isnull().sum())

Missing Values Report
--------------------------------------------------

orders
order_id                0
customer_id             0
order_date              0
order_time              0
status                  0
city                    0
state                   0
pincode                 0
total_amount            0
gst_amount              0
shipping_charge         0
discount_amount         0
final_amount            0
payment_method          0
shipping_partner        0
tracking_id             0
delivered_date      17501
is_cod                  0
channel                 0
dtype: int64

order_items
item_id            0
order_id           0
product_id         0
product_name       0
category           0
quantity           0
unit_price         0
gst_rate           0
gst_amount         0
discount_amount    0
total_price        0
dtype: int64

suppliers
supplier_id           0
supplier_name         0
contact_person        0
email                 0
phone                 0
city                  0


In [ ]:
print("---Missing values Analysis ----")

for name,df in datasets.items():
  null_counts=df.isnull().sum()
  null_pct=(df.isnull().sum()/len(df))*100

  missing_info=pd.DataFrame({
      'Missing_Count':null_counts,
      'missing_pct(%)':null_pct
  })
  only_missing=missing_info[missing_info['Missing_Count']>0]
  print(f"\n--- Table :{name.upper()}----")


  if missing_info.empty:
     print("✅ No missing values found.")
  else:
     print("Missing values found:")
     print(missing_info)

---Missing values Analysis ----

--- Table :ORDERS----
Missing values found:
                  Missing_Count  missing_pct(%)
order_id                      0           0.000
customer_id                   0           0.000
order_date                    0           0.000
order_time                    0           0.000
status                        0           0.000
city                          0           0.000
state                         0           0.000
pincode                       0           0.000
total_amount                  0           0.000
gst_amount                    0           0.000
shipping_charge               0           0.000
discount_amount               0           0.000
final_amount                  0           0.000
payment_method                0           0.000
shipping_partner              0           0.000
tracking_id                   0           0.000
delivered_date            17501          35.002
is_cod                        0           0.000
channel    

**Step 4: Check for duplicates rows across all datasets**

In [ ]:
# Duplicate summary
duplicate_summary = []

for name, df in datasets.items():
    duplicate_summary.append({
        "Dataset": name,
        "Total Rows": len(df),
        "Duplicate Rows": df.duplicated().sum()
    })

duplicate_df = pd.DataFrame(duplicate_summary)

print(duplicate_df)

       Dataset  Total Rows  Duplicate Rows
0       orders       50000               0
1  order_items      100000               0
2    suppliers         200               0
3    inventory        1000               0
4     payments       50000               0
5    customers       10000               0
6     products        1000               0
7      returns       10000               0


**Step 5: Convert Date columns to Proper Datetime Objects across all tables**

In [ ]:
orders_df ['order_date'] = pd.to_datetime(orders_df['order_date'], format="%d-%m-%Y", errors='coerce')
orders_df['delivered_date'] = pd.to_datetime(orders_df['delivered_date'], format="%d-%m-%Y", errors='coerce')
customers_df['registration_date'] = pd.to_datetime(customers_df['registration_date'], format = "%d-%m-%Y", errors='coerce')
payments_df['payment_date'] = pd.to_datetime(payments_df['payment_date'], format="%d-%m-%Y", errors='coerce')

print("=== ALL DATE COLUMNS SUCCESSFULLY CONVERTED ===")
print("orders_df -> order_date type:", orders_df['order_date'].dtype)
print("customers_df -> registration_date type:", customers_df['registration_date'].dtype)


=== ALL DATE COLUMNS SUCCESSFULLY CONVERTED ===
orders_df -> order_date type: datetime64[ns]
customers_df -> registration_date type: datetime64[ns]


**Step 6: Check For Outliers in Final Amount (>5 Lakhs)**

In [ ]:
outliers=orders_df [orders_df ['final_amount'] >500000]

print('=== OUTLIER ANALYSIS (> 5 Lakhs) ===')

print(f"number of orders above ₹500,000: {len (outliers)}")



if len(outliers) > 0:

    print("\nOutlier Details:")

    print(outliers [['order_id', 'customer_id', 'final_amount', 'status']])
else:

    print("No extreme financial outliers detected above 5 Lakhs.")

=== OUTLIER ANALYSIS (> 5 Lakhs) ===
number of orders above ₹500,000: 312

Outlier Details:
        order_id customer_id  final_amount     status
33     ORD000034   CUST01867     527178.05  Delivered
34     ORD000035   CUST01370     532061.13  Cancelled
306    ORD000307   CUST03874     550716.17  Cancelled
583    ORD000584   CUST07624     745667.68  Delivered
590    ORD000591   CUST05298     520293.72  Cancelled
...          ...         ...           ...        ...
39194  ORD039195   CUST02704     588456.74  Delivered
39582  ORD039583   CUST06723     519500.64  Delivered
39793  ORD039794   CUST04753     640296.82  Delivered
39804  ORD039805   CUST04071     796481.05  Delivered
39888  ORD039889   CUST03676     537504.43  Cancelled

[312 rows x 4 columns]


**Step 7:Specific Business Logic Checks**

In [ ]:
print("=== SPECIFIC BUSINESS QUALITY CHECKS ===")

#1. Invalid Delivery Dates Check

invalid_delivery = orders_df [orders_df['delivered_date'] < orders_df['order_date']]

print(f"1. Orders with delivered_date BEFORE order_date: {len(invalid_delivery)}")

#2. Payment Gateway Failure Rate


status_col='payment_status' if 'payment_status' in payments_df.columns else 'status'

failed_payments = payments_df [payments_df [status_col] == 'Failed']

failure_rate = (len(failed_payments) / len(payments_df)) * 100

print(f"2. Payment Failure Rate: {failure_rate:.2f}% ({len(failed_payments)} failed out of {len(payments_df)})")



#3. Returns Referential Integrity Check

unmatched_returns=returns_df [~returns_df ['order_id'].isin(orders_df ['order_id'])]

print(f"3. Returns with no matching order_id in orders.csv: {len(unmatched_returns)}")

#4. Inventory Out-of-Stock Check

out_of_stock=inventory_df[inventory_df['status'] == 'Out of Stock']

print(f"4. Products in 'Out of Stock' status: {len(out_of_stock)}")

#5. Customers with Zero Orders Check

zero_order_cust=customers_df[customers_df ['total_orders'] == 0]

print(f"5. Registered customers with e total orders: {len (zero_order_cust)}")

=== SPECIFIC BUSINESS QUALITY CHECKS ===
1. Orders with delivered_date BEFORE order_date: 0
2. Payment Failure Rate: 3.50% (1748 failed out of 50000)
3. Returns with no matching order_id in orders.csv: 0
4. Products in 'Out of Stock' status: 2
5. Registered customers with e total orders: 457


#Phase 1: Data Quality Audit Report
##1. Overview & Dataset Profile
This report summarizes the initial data quality assessment performed on the provided datasets: `orders`, `order_items`, `suppliers`, `inventory`, `payments`, `customers`, `products`, and `returns`.
#
#
* Primary key integrity: confirmed 0 duplicate `order_id` values in `orders.csv` and 0 duplicate rows across all tables

## 1. Data Loading and Overview
All 8 datasets were successfully loaded into Pandas DataFrames. Below is a summary of their dimensions:

| Dataset     | Total Rows | Total Columns |
|-------------|------------|---------------|
| orders      | 50000      | 19            |
| order_items | 100000     | 11            |
| suppliers   | 200        | 14            |
| inventory   | 1000       | 11            |
| payments    | 50000      | 13            |
| customers   | 10000      | 17            |
| products    | 1000       | 17            |
| returns     | 10000      | 10            |

## 2. Duplicate Analysis
No duplicate rows were found across any of the datasets, indicating good data integrity in terms of unique records.

## 3. Missing Values Analysis
Missing values were identified in two datasets:

*   **orders_df**: The `delivered_date` column has 17,501 missing values, accounting for approximately 35.002% of the records. This indicates a significant portion of orders where the delivery date is not recorded.
*   **payments_df**: The `refund_date` column has 50,000 missing values, which is 100% of its entries. This suggests that refunds are either not tracked, or the column is intended to be null for non-refunded payments.

Other datasets (order_items, suppliers, inventory, customers, products, returns) have no missing values.

## 4. Data Type Conversion
Date columns in `orders_df`, `customers_df`, and `payments_df` (`order_date`, `delivered_date`, `registration_date`, `payment_date`) were successfully converted to `datetime64[ns]` format. This is crucial for time-series analysis and accurate date comparisons.

## 5. Outlier Analysis (Final Amount > ₹500,000)
An outlier analysis on `orders_df` for `final_amount` revealed 312 orders with `final_amount` exceeding ₹500,000. These outliers might represent high-value orders or potential data entry errors and warrant further investigation.

## 6. Specific Business Quality Checks
Several business-specific quality checks were performed:

*   **Invalid Delivery Dates**: No orders were found where `delivered_date` was prior to `order_date`, indicating consistency in date logic.
*   **Payment Gateway Failure Rate**: The payment failure rate was calculated at **3.50%** (1,748 failed payments out of 50,000). This metric could be an area for operational improvement.
*   **Returns Referential Integrity**: All `order_id`s in the `returns_df` have a matching `order_id` in `orders_df`, confirming good referential integrity for returns data.
*   **Inventory Out-of-Stock**: 2 products were found to be 'Out of Stock' in the `inventory_df`. This highlights specific inventory management issues.
*   **Customers with Zero Orders**: 457 registered customers have a `total_orders` count of 0. This segment could be targeted for engagement campaigns or re-evaluation of registration processes.

## Summary and Recommendations
Overall, the datasets demonstrate good quality with no duplicates and successful data type conversions. However, significant missing values in `orders_df['delivered_date']` and `payments_df['refund_date']` require attention. The presence of high-value order outliers and a 3.50% payment failure rate also suggests areas for further business investigation and potential data cleaning or process improvements.